### PR TITLE
use `casted_rhs` instead of `rhs` so `icmp` works correctly for `airShlSat`

### PR DESCRIPTION
**Pull request by stefansu28**
Closes https://github.com/ziglang/zig/issues/18302

The generated `in_range` llvm value was based on the original `rhs` which could be of the wrong type. Using `casted_rhs` resolves this issue.

### CODE REVIEW

**Comment by matu3ba** [comment]
Not sure, if the test case belongs into test\cases\ or test\behavior. Currently it is missing.

**Comment by stefansu28** [comment]
I added a test the behavior tests, but it looks like some of the builds are sad since saturating shift left isn't implemented.
<details>
<summary>error</summary>
<pre>
> stage4/bin/zig build test-behavior -Dskip-release -Dskip-non-native
zig test Debug x86_64-linux-none: error: the following command failed with 1 compilation errors:
/home/stefan/code/zig/build/stage4/bin/zig test -fno-llvm -fno-lld /home/stefan/code/zig/test/behavior.zig --cache-dir /home/stefan/code/zig/zig-cache --global-cache-dir /home/stefan/.cache/zig --name test -target x86_64-linux-none -mcpu x86_64 -I /home/stefan/code/zig/test --zig-lib-dir /home/stefan/code/zig/lib --listen=- 
zig test Debug x86_64-linux-none: error: the following command failed with 1 compilation errors:
/home/stefan/code/zig/build/stage4/bin/zig test -fno-llvm -fno-lld /home/stefan/code/zig/test/behavior.zig -fstrip --cache-dir /home/stefan/code/zig/zig-cache --global-cache-dir /home/stefan/.cache/zig --name test -target x86_64-linux-none -mcpu x86_64_v3 -I /home/stefan/code/zig/test --zig-lib-dir /home/stefan/code/zig/lib --listen=- 
zig test Debug x86_64-linux-none: error: the following command failed with 1 compilation errors:
/home/stefan/code/zig/build/stage4/bin/zig test -fno-llvm -fno-lld /home/stefan/code/zig/test/behavior.zig --cache-dir /home/stefan/code/zig/zig-cache --global-cache-dir /home/stefan/.cache/zig --name test -target x86_64-linux-none -mcpu x86_64_v2 -I /home/stefan/code/zig/test --zig-lib-dir /home/stefan/code/zig/lib -fPIC --listen=- 
Build Summary: 16/24 steps succeeded; 1 skipped; 3 failed; 10843/11447 tests passed; 604 skipped (disable with --summary none)
test-behavior transitive failure
├─ run test behavior-x86_64-linux-none-x86_64-Debug-selfhosted-no-lld transitive failure
│  └─ zig test Debug x86_64-linux-none 1 errors
├─ run test behavior-x86_64-linux-none-x86_64_v2-Debug-selfhosted-no-lld-pic transitive failure
│  └─ zig test Debug x86_64-linux-none 1 errors
├─ run test behavior-x86_64-linux-none-x86_64_v3-Debug-selfhosted-no-lld transitive failure
│  └─ zig test Debug x86_64-linux-none 1 errors
└─ run test behavior-x86_64-linux-gnu-x86_64-Debug-libc skipped
/home/stefan/code/zig/test/behavior/bugs/18302.zig:24:13: error: TODO implement shl_sat for target.Target.Cpu.Arch.x86_64
        pub fn shiftExponent(self: @This(), shift: ShiftType) @This() {
        ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/stefan/code/zig/test/behavior/bugs/18302.zig:24:13: error: TODO implement shl_sat for target.Target.Cpu.Arch.x86_64
        pub fn shiftExponent(self: @This(), shift: ShiftType) @This() {
        ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/stefan/code/zig/test/behavior/bugs/18302.zig:24:13: error: TODO implement shl_sat for target.Target.Cpu.Arch.x86_64
        pub fn shiftExponent(self: @This(), shift: ShiftType) @This() {
        ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
</pre>
</details>

Not sure if I should put this test somewhere else or if there is some way to skip this test for that build.

File: test/behavior.zig [hunk 1]
```diff
@@ -143,6 +143,7 @@ test {
     _ = @import("behavior/bugs/13785.zig");
     _ = @import("behavior/bugs/14854.zig");
     _ = @import("behavior/bugs/15778.zig");
+    _ = @import("behavior/bugs/18302.zig");
```
Comment (Vexu) [inline test/behavior.zig:146]: Do not add new `bugs/XXXX.zig` tests, instead place it one of the categorized files.

**Review by Vexu — CHANGES_REQUESTED** [review]
> Not sure if I should put this test somewhere else or if there is some way to skip this test for that build.

You can skip the test by checking the target and returning `error.SkipZigTest`:
```zig
    if (builtin.zig_backend == .stage2_x86_64) return error.SkipZigTest; // TODO
```

File: test/behavior.zig (hunk 1 above)
Comment (stefansu28) [reply test/behavior.zig]: Moved the test to `bit_shifting.zig`

**Review by Vexu — CHANGES_REQUESTED** [review]
Could you check the result in the test? And that helper function looks like a reimplementation of `std.math.Log2Int`.

**Comment by stefansu28** [comment]
`std.math.Log2Int` did not reproduce the error on master, but converting the type to `.signed` did.

**Pull request merged**